src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Analyzer.cs

```diff
@@ -0,0 +1,238 @@
+                return;
+            }
+
+            // IndexOf(string, int startIndex)
+            IMethodSymbol? indexOfStringInt32Method = indexOfMethods.GetFirstOrDefaultMemberWithParameterInfos(stringInt32Parameters);
+            if (indexOfStringInt32Method == null)
+            {
+                return;
+            }
+
+            // IndexOf(string, int startIndex, int count)
+            IMethodSymbol? indexOfStringInt32Int32Method = indexOfMethods.GetFirstOrDefaultMemberWithParameterInfos(stringInt32Int32Parameters);
+            if (indexOfStringInt32Int32Method == null)
+            {
+                return;
+            }
+
+            // CompareTo(string)
+            IMethodSymbol? compareToStringMethod = stringType.GetMembers(StringCompareToMethodName).OfType<IMethodSymbol>().GetFirstOrDefaultMemberWithParameterInfos(stringParameter);
+            if (compareToStringMethod == null)
+            {
+                return;
+            }
+
+            // Retrieve the StringComparer properties that need to be flagged: CurrentCultureIgnoreCase, InvariantCultureIgnoreCase
+
+            IEnumerable<IPropertySymbol> ccicPropertyGroup = stringComparerType.GetMembers(StringComparisonCurrentCultureIgnoreCaseName).OfType<IPropertySymbol>();
+            if (!ccicPropertyGroup.Any())
+            {
+                return;
+            }
+
+            IEnumerable<IPropertySymbol> icicPropertyGroup = stringComparerType.GetMembers(StringComparisonInvariantCultureIgnoreCaseName).OfType<IPropertySymbol>();
+            if (!icicPropertyGroup.Any())
+            {
+                return;
+            }
+
+            context.RegisterOperationAction(context =>
+            {
+                IInvocationOperation caseChangingInvocation = (IInvocationOperation)context.Operation;
+                IMethodSymbol caseChangingMethod = caseChangingInvocation.TargetMethod;
+
+                if (!caseChangingMethod.Equals(toLowerParameterlessMethod) &&
+                    !caseChangingMethod.Equals(toLowerInvariantParameterlessMethod) &&
+                    !caseChangingMethod.Equals(toUpperParameterlessMethod) &&
+                    !caseChangingMethod.Equals(toUpperInvariantParameterlessMethod))
+                {
+                    return;
+                }
+
+                // Ignore parenthesized operations
+                IOperation? ancestor = caseChangingInvocation.WalkUpParentheses().WalkUpConversion().Parent;
+
+                IInvocationOperation diagnosableInvocation;
+                if (ancestor is IInvocationOperation invocationAncestor)
+                {
+                    diagnosableInvocation = invocationAncestor;
+                }
+                else if (ancestor is IArgumentOperation argumentAncestor && argumentAncestor.Parent is IInvocationOperation argumentInvocationAncestor)
+                {
+                    diagnosableInvocation = argumentInvocationAncestor;
+                }
+                else
```

Comment:
Should we consider detecting simple patterns across statements, such as below?

```csharp
var loweredStr = str.ToLower();
if (loweredStr.Contains(anotherStr))
```

We would need to check that we have exactly one reference to the temp variable. If you are interested, we already added another analyzer which implements a simple heuristic for above patterns: https://github.com/dotnet/roslyn-analyzers/blob/main/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOfAnalyzer.cs. If not, you may also feel free to also file a tracking issue for this enhancement in future if you deem it to be worthwhile.

Reply:
Yes, V2. Opened: https://github.com/dotnet/roslyn-analyzers/issues/6689